Add integration for offline support

voraciousdev opened the pull request:
Before submitting a pull request, please take a look at our
[Contributing](https://github.com/getsentry/sentry-javascript/blob/master/CONTRIBUTING.md) guidelines and verify:

- [x] If you've added code that should be tested, please add tests.
- [x] Ensure your code lints and the test suite passes (`yarn lint`) & (`yarn test`).

### Summary

Cache offline events to send when connectivity is reestablished.

This PR is an adaptation of #2216. Resolves #1633.

voraciousdev commented:
Okay, I think I have this in a good place now, but I still need to write some tests. The final failing check is related to bundle size, so I'm not entirely sure how to proceed. I assume this is from the addition of the `localforage` dependency.

AbhiPrasad commented:
This looks good at an initial glance, but before we do a deeper review, @voraciousdev do you mind if you could move this integration into our integrations package? I think this is the best place for features like this.

https://github.com/getsentry/sentry-javascript/tree/master/packages/integrations.

Also do you have a demo repo or something of the sort you've been testing this with? Would help with review/testing on our side (also to explore all the use cases).

Thanks for taking this on, the Sentry team appreciates it ❤️

voraciousdev commented:
@AbhiPrasad I appreciate the feedback! I believe I moved this to the correct place now. The repo I've been using to test this is closed source, but I'll see if I can throw an example together today.

AbhiPrasad commented on packages/integrations/package.json:
```diff
@@ -16,8 +16,10 @@
   "module": "esm/index.js",
   "types": "dist/index.d.ts",
   "dependencies": {
+    "@sentry/minimal": "5.20.1",
```
I don't think we need the `@sentry/minimal` depedency.

Since we are in an integration, we have access to the hub with `getCurrentHub()`. This means we can use `hub.captureEvent()` - https://getsentry.github.io/sentry-javascript/classes/hub.hub-1.html#captureevent

AbhiPrasad commented on packages/integrations/src/offline.ts:
```diff
@@ -0,0 +1,101 @@
+import { Event, EventProcessor, Hub, Integration } from '@sentry/types';
+import { getGlobalObject, logger, uuid4 } from '@sentry/utils';
+// @ts-ignore
+import localforage = require('localforage');
```
Any reason we are not using import statement here?

voraciousdev replied:
😄  There sure is... I'll see if I can link some open issues on localforage, but the gist is this: localforage does not export modules in a way compatible with this library. I attempted each of the suggested fixes (via the README and related issues), but I believe there is actually a bug in the package.

AbhiPrasad replied:
Huh interesting. I will take a look, maybe it is a problem with our rollbar/bundling setup.

voraciousdev replied:
It's definitely possible that I just missed something, but I tried quite a few suggestions. The issues I ran into can be reproduced by following the [suggestions in the localforage README for typescript](https://github.com/localForage/localForage#typescript).

voraciousdev replied:
I'm going to mark this as resolved, because I ended up figuring out what was wrong.

AbhiPrasad commented on packages/integrations/src/offline.ts:
```diff
@@ -0,0 +1,101 @@
+   * @inheritDoc
+   */
+  public setupOnce(addGlobalEventProcessor: (callback: EventProcessor) => void, getCurrentHub: () => Hub): void {
+    addGlobalEventProcessor(async (event: Event) => {
```
We try to avoid using `async` in non-test codebase as it can have much higher bundle size vs. just using regular Promises.

voraciousdev replied:
Ah, got it. I can convert it.

voraciousdev replied:
Okay, I removed the `async/await` calls.

AbhiPrasad commented on packages/integrations/src/offline.ts:
```diff
@@ -0,0 +1,101 @@
+   * @inheritDoc
+   */
+  public constructor() {
+    this.offlineEventStore = localforage.createInstance({
```
Could `createInstance` fail? What happens to `this.offlineEventStore` then?

We can import localforage typings if they exist as well.

Also, what browsers does `localforage` support? It's fine if it doesn't match Sentry's (https://docs.sentry.io/platforms/javascript/#browser-table) - we just have to document it then.

voraciousdev replied:
Ah, good catch. I meant to wrap this in a `try/catch`. I will look into importing the type file too.

voraciousdev replied:
I think it meets the compatibility requirements. It is a wrapper around multiple storage mechanisms. It attempts to use IndexedDB, then falls back to WebSQL, and if all else fails, it falls back to localStorage.
https://github.com/localForage/localForage/wiki/Supported-Browsers-Platforms

AbhiPrasad replied:
Nice ok, makes sense. Just one last thing, what happens when local storage is full? Is it a queue, or do new events just get dropped?

@bruno-garcia, how do we handle offline queue in mobile right now?

bruno-garcia replied:
Only envelopes are cached. The order they were cached is kept when attempting to send things to Sentry (i.e device is back online). If `maxItems` is reached (by default 30 IIRC), the oldest file is deleted to make space to the newest one.

We don't use dependencies or a DB of any sort. It's just files written to a directory.

If session data is stored (release health), the `init=true` flag must be preserved. So when deleting a file to make room for a new one (max items reached), the envelope would need to be unwrapped, checked for sessions and if init=true exists, that must be moved to the next session update queued up for submission (from oldest to newest).
Events capture when session tracking is on must be in an envelope together with a session update.

Note that the order being kept means that if you get a call to `captureException`, you won't be sending that to Sentry if there's stuff cached. You must prioritize what's on the storage first.

AbhiPrasad replied:
Thanks Bruno for coming through 🚀. We can ignore all the sessions stuff, and envelope details - although important for me, probably not as for you @voraciousdev

Basically there are two things we need to keep in mind.
1) Enforce some kind of `maxItems` in local storage. This can be changed through an option.

2) Make sure that we send all events from local storage first before we continue on with other events. I know we are using an event listener right now to listen to `online`, do you think that is good enough? Maybe we need another check in the `globalEventProcessor` to be sure.

As for how this will work with sessions, we can get to that when release health comes to JS 😄

voraciousdev replied:
@AbhiPrasad the `setItem` promise will be rejected if there is not enough space on the device. With the way it's currently written in this integration, that would result in `logger.warn('could not cache event while offline')` being called.

AbhiPrasad replied:
Ok perfect we should we good then.

voraciousdev replied:
> Make sure that we send all events from local storage first before we continue on with other events. I know we are using an event listener right now to listen to `online`, do you think that is good enough? Maybe we need another check in the `globalEventProcessor` to be sure.

@AbhiPrasad I'm not sure if this is directed at me, but there is definitely a possible race condition here regarding the `online` event. Given the nature of this integration (to support offline-capable apps), I think it's a real scenario that we could have real error events triggered as soon as the `online` event is fired. For offline apps, it's common practice to listen to the `online` event and run some code (which could then lead to an error being thrown/recorded).

voraciousdev replied:
Just added localforage typings as well. Good call on that. 👍

voraciousdev commented:
@AbhiPrasad I believe I have addressed all open comments, and I just added tests. Mind taking another look and letting me know if there's anything else you need me to change? Thanks!

voraciousdev commented:
@AbhiPrasad I did some more testing in my own project, and I discovered a couple of small issues that I believe are now fixed via the extra commits you see here. To break them down:

- Fixing lockfile after migrating to @sentry/integrations
  - It turns out I had some lockfile changes that were committed when this was initially added to the `@sentry/browser` package.
- Add TS compiler flag to allow import of localforage
  - The import assignment was not working properly in my es6/webpack project. It looks like some of the builds were failing. After some more debugging, it looks like using a standard import (e.g. `import localforage from 'localforage'`) actually _does_ work. It just throws a typescript error due to the atypical nature of the package exports.
- Handle scenario where app has stored events from previous offline session
  - This is just a new scenario that I didn't previously think about. When an app is offline, just listening to the `online` event is not good enough. If the user navigates away from the app while offline and then navigates back while online, the `online` event would never be fired. Because of this, we will want to check for connectivity during the `setupOnce` call and send any previously stored events.

voraciousdev commented:
@HazAT thank you for following up! I can definitely manage that. I'll tag you for another review once it's ready.

HazAT commented on packages/integrations/package.json:
```diff
@@ -18,6 +18,7 @@
   "dependencies": {
     "@sentry/types": "5.20.1",
     "@sentry/utils": "5.20.1",
```
```suggestion
    "@sentry/types": "5.21.3",
    "@sentry/utils": "5.21.3",
```

voraciousdev replied:
It looks like those deps were just updated to `5.21.4`, so I used that version instead!

HazAT reviewed (changes requested):
Hey @voraciousdev thanks for keeping the ball rolling, aside from the merge conflicts I think we are good to go :)

voraciousdev commented:
@HazAT it looks like there were some lint rules that changed since I started this. I will get it updated.

voraciousdev commented:
I don't usually like merge commits, but the GitHub UI created one automatically when I resolved those conflicts. Let me know if you would prefer a rebase.

HazAT reviewed (approved):
We squash commit anyway :) 
Thanks @voraciousdev contribution of the year 🥇

voraciousdev commented:
@HazAT any idea when this will be released?

kamilogorek commented:
@voraciousdev tomorrow :)

piotr-cz commented on packages/integrations/src/offline.ts:
```diff
@@ -0,0 +1,169 @@
+        if (this.hub) {
+          const newEventId = this.hub.captureEvent(event);
+
+          if (newEventId) {
```
The `Hub.captureEvent` method [always returns new event ID](https://github.com/getsentry/sentry-javascript/blob/2e761fe183a7e6d63fd438dc0aa9f071c2fff5ea/packages/hub/src/hub.ts#L210), so event will always be purged from event store (even that client failed to send it).

piotr-cz replied:
Fixed in #2890

piotr-cz commented on packages/integrations/src/offline.ts:
```diff
@@ -0,0 +1,169 @@
+      name: 'sentry/offlineEventStore',
+    });
+
+    if ('addEventListener' in this.global) {
```
Event listener should be set up _after_ hub is set on instance (in `setupOnce` method) as event may be fired before hub is set (hub is used in `_sendEvents`).

IMHO it's bad practice to have business logic in constructor.

piotr-cz replied:
Fixed in #2889
